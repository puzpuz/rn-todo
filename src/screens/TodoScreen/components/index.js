import AddButton from "./AddButton";
import TodoItem from "./TodoItem";

export { AddButton, TodoItem };
