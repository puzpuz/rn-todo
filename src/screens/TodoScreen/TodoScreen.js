import React, { useCallback, useEffect, useState } from "react";
import {
  FlatList,
  KeyboardAvoidingView,
  Platform,
  Text,
  TextInput,
  View,
} from "react-native";
import uuid from "react-native-uuid";
import { AddButton } from "./components";
import TodoItem from "./components/TodoItem";
import styles from "./styles/todoScreenStyles";

const INITIAL_TODO = [
  {
    id: uuid.v4(),
    value: "Todo 1",
  },
  {
    id: uuid.v4(),
    value: "Todo 2",
  },
];

const TodoScreen = ({ navigation }) => {
  const [todoList, setTodoList] = useState();
  const [addInput, setAddInput] = useState("");
  const [updateItem, setUpdateItem] = useState();

  //Reseting text input after submit
  const resetInput = () => {
    setAddInput("");
  };

  //handle add and update todolist state array
  const onAddTodo = () => {
    if (addInput === "") return;

    //if updateItem is empty then add new item at the bottom of the array
    if (!updateItem) {
      setTodoList([
        ...todoList,
        {
          id: uuid.v4(),
          value: addInput,
        },
      ]);
      resetInput();
      //if updateItem not empty, find the updated item index, and replace with new value
    } else {
      const index = todoList.findIndex((todo) => todo.id === updateItem.id);
      console.debug(index, updateItem);
      if (index && index >= 0) {
        setTodoList([
          ...todoList.slice(0, index),
          { ...updateItem, value: addInput },
          ...todoList.slice(index + 1),
        ]);
        resetInput();
        setUpdateItem(undefined);
      }
    }
  };

  //handle remove an item from todo list state array
  const onRemoveTodo = (item) => {
    const index = todoList.findIndex((todo) => todo.id === item.id);
    if (index && index >= 0) {
      setTodoList([...todoList.slice(0, index), ...todoList.slice(index + 1)]);
    }
  };

  //Logic to asssign update item
  const onSetUpdateItem = (item) => {
    setUpdateItem(item);
    setAddInput(item.value);
  };

  const renderTodoItem = ({ item, index }) => (
    <TodoItem
      item={item}
      onRemoveTodo={onRemoveTodo}
      onSetUpdateItem={onSetUpdateItem}
    />
  );

  const todoKeyExtractor = useCallback((item, index) => `todo${index}`, []);

  useEffect(() => {
    setTodoList(INITIAL_TODO);
  }, []);

  return (
    <KeyboardAvoidingView
      style={styles.todoContainer}
      behavior={Platform.OS === "ios" ? "padding" : "height"}
    >
      <Text style={styles.title}>Todo:</Text>
      <FlatList
        style={styles.flatListContainer}
        data={todoList}
        renderItem={renderTodoItem}
        keyExtractor={todoKeyExtractor}
      />
      <View style={styles.footerContainer}>
        <TextInput
          style={styles.input}
          placeholder="Add new todos"
          onChangeText={(value) => setAddInput(value)}
          value={addInput}
        />
        <AddButton onAddTodo={onAddTodo} updateItem={updateItem} />
      </View>
    </KeyboardAvoidingView>
  );
};

export default TodoScreen;
