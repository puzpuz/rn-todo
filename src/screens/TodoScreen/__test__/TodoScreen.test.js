import renderer from "react-test-renderer";
import TodoScreen from "../TodoScreen";
it("renders correctly", () => {
  const tree = renderer.create(<TodoScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
