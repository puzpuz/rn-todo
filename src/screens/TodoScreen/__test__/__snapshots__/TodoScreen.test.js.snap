// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "alignItems": "flexStart",
        "backgroundColor": "#f0f0f0",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "flexStart",
        "paddingHorizontal": 8,
        "paddingVertical": 8,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <Text
    style={
      {
        "color": "blue",
        "fontSize": 32,
        "marginVertical": 32,
      }
    }
  >
    Todo:
  </Text>
  <RCTScrollView
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    style={
      {
        "width": "100%",
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "borderRadius": 24,
        "flexDirection": "row",
        "justifyContent": "center",
        "paddingHorizontal": 8,
        "paddingVertical": 16,
        "width": "100%",
      }
    }
  >
    <TextInput
      onChangeText={[Function]}
      placeholder="Add new todos"
      style={
        {
          "flex": 1,
          "fontSize": 18,
        }
      }
      value=""
    />
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          {
            "backgroundColor": "blue",
            "borderRadius": 24,
            "padding": 24,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          ADD
        </Text>
      </View>
    </View>
  </View>
</View>
`;
